AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/DefaultKeyStorageInterface.dfy

```diff
@@ -708,6 +773,522 @@ module DefaultKeyStorageInterface {
+      requires
+        && ValidState()
+      modifies Modifies - {History}
+      // Dafny will skip type parameters when generating a default decreases clause.
+      decreases Modifies - {History}
+      ensures
+        && ValidState()
+      ensures WriteInitializeMutationEnsuresPublicly(input, output)
+      ensures unchanged(History)
+    {
+
+      :- Need(
+        Structure.MutationLock?(input.mutationLock)
+      , Types.KeyStorageException(
+          message := "Invalid mutation lock."
+        )
+      );
+
+        /** Validate Inputs can be mapped to DDB Items */
+      :- Need(
+        && (forall k <- input.version.EncryptionContext.Keys :: DDB.IsValid_AttributeName(k))
+        && (forall k <- input.active.EncryptionContext.Keys :: DDB.IsValid_AttributeName(k))
+        && (forall k <- input.beacon.EncryptionContext.Keys :: DDB.IsValid_AttributeName(k))
+      , Types.KeyStorageException( message := ErrorMessages.ENCRYPTION_CONTEXT_EXCEEDS_DDB_LIMIT)
+      );
+      /** Convert Inputs to DDB Items.*/
+      var items: DDB.TransactWriteItemList := [
+        CreateTransactWritePutItem(
+          input.version,
+          ddbTableName,
+          BRANCH_KEY_NOT_EXIST // The new Decrypt Only MUST not exist
+        ),
+        CreateTransactOverwriteActive(
+          input.active,
+          input.oldActive,
+          ddbTableName
+        ),
+        CreateTransactWritePutItem(
+          input.beacon,
+          ddbTableName,
+          BRANCH_KEY_EXISTS // The Beacon MUST exist
+        ),
+        CreateTransactWriteMLock(
+          input.mutationLock,
+          ddbTableName,
+          BRANCH_KEY_NOT_EXIST // The M-LOCK MUST not exist
+        )
+      ];
+      var transactRequest := DDB.TransactWriteItemsInput(
+        TransactItems := items,
+        ReturnConsumedCapacity := None,
+        ReturnItemCollectionMetrics := None,
+        ClientRequestToken := None
+      );
+
+      var transactWriteItemsResponse? := ddbClient.TransactWriteItems(transactRequest);
+      // TODO: Post Beta, we need to check the cancellation reason for
+      // ConditionalCheckFailed on the Active item (VersionRaceException)
+      // OR the Mutation Lock (MutationLockException)
+      // OR something else.
+      var _ :- transactWriteItemsResponse?
+      .MapFailure(e => wrapDdbException(
+                      e:=e,
+                      storageOperation:="WriteInitializeMutation",
+                      ddbOperation:="TransactWriteItems",
+                      identifier:=input.active.Identifier,
+                      tableName:=ddbTableName));
+      output := Success(Types.WriteInitializeMutationOutput());
+    }
+
+    static const queryForVersionsKeyExpression: DDB.KeyExpression := "#pk = :pk AND begins_with( #sk, :decryptOnlyPrefix )"
+    static const queryForVersionsExpAttNames: DDB.ExpressionAttributeNameMap :=
+      map[
+        "#pk" := Structure.BRANCH_KEY_IDENTIFIER_FIELD,
+        "#sk" := Structure.TYPE_FIELD]
```

Comment:
> add all the consts in this file to a shared file? Structure.dfy maybe?

These constants ONLY apply to DynamoDB.
Technically, it is a mistake that Structure.dfy does not fully abstract the DynamoDB types to this module.
i.e: Structure requires DynamoDB inputs.
Thus, non-DynamoDB storage MUST implement the logic in Structure.

I gave this feedback on the Storage PR.
But it is not a high priority.

As for defining them at the top of file:
I understand why developers define constants at the top of files.
Constants that will be used in may methods/functions of a logical code group 
should be defined in a common spot.

However,
that is not why I have created these constants.
They are ONLY used by `QueryForVersions`.

They have no other purpose;
nor can they really have any other purpose.

Therefore, 
I believe it is actually better to have them defined here.
The only reason they are defined as constants is because they are static;
it is a tiny optimization to create these strings at class loading time,
as compared to in the function/methods runtime.

Moving them to the top of the file just puts physical distance 
between their definition and their one and only use...